File: fluss-rust/bindings/python/test/test_sasl_auth.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

"""Integration tests for SASL/PLAIN authentication.

Mirrors the Rust integration tests in crates/fluss/tests/integration/sasl_auth.rs.
"""

import pytest

import fluss


async def test_sasl_connect_with_valid_credentials(sasl_bootstrap_servers):
    """Verify that a client with correct SASL credentials can connect and perform operations."""
    config = fluss.Config({
        "bootstrap.servers": sasl_bootstrap_servers,
        "security.protocol": "sasl",
        "security.sasl.mechanism": "PLAIN",
        "security.sasl.username": "admin",
        "security.sasl.password": "admin-secret",
    })
    conn = await fluss.FlussConnection.create(config)
    admin = conn.get_admin()

    db_name = "py_sasl_test_valid_db"
    db_descriptor = fluss.DatabaseDescriptor(comment="created via SASL auth")
    await admin.create_database(db_name, db_descriptor, ignore_if_exists=True)

    assert await admin.database_exists(db_name)

    # Cleanup
    await admin.drop_database(db_name, ignore_if_not_exists=True, cascade=True)
    await conn.close()


async def test_sasl_connect_with_second_user(sasl_bootstrap_servers):
    """Verify that a second user can also authenticate successfully."""
    config = fluss.Config({
        "bootstrap.servers": sasl_bootstrap_servers,
        "security.protocol": "sasl",
        "security.sasl.mechanism": "PLAIN",
        "security.sasl.username": "alice",
        "security.sasl.password": "alice-secret",
    })
    conn = await fluss.FlussConnection.create(config)
    admin = conn.get_admin()

    # Basic operation to confirm functional connection
    assert not await admin.database_exists("some_nonexistent_db_alice")
    await conn.close()


async def test_sasl_connect_with_wrong_password(sasl_bootstrap_servers):
    """Verify that wrong credentials are rejected with AUTHENTICATE_EXCEPTION."""
    config = fluss.Config({
        "bootstrap.servers": sasl_bootstrap_servers,
        "security.protocol": "sasl",
        "security.sasl.mechanism": "PLAIN",
        "security.sasl.username": "admin",
        "security.sasl.password": "wrong-password",
    })
    with pytest.raises(fluss.FlussError) as exc_info:
        await fluss.FlussConnection.create(config)

    assert exc_info.value.error_code == fluss.ErrorCode.AUTHENTICATE_EXCEPTION


async def test_sasl_connect_with_unknown_user(sasl_bootstrap_servers):
    """Verify that a nonexistent user is rejected with AUTHENTICATE_EXCEPTION."""
    config = fluss.Config({
        "bootstrap.servers": sasl_bootstrap_servers,
        "security.protocol": "sasl",
        "security.sasl.mechanism": "PLAIN",
        "security.sasl.username": "nonexistent_user",
        "security.sasl.password": "some-password",
    })
    with pytest.raises(fluss.FlussError) as exc_info:
        await fluss.FlussConnection.create(config)

    assert exc_info.value.error_code == fluss.ErrorCode.AUTHENTICATE_EXCEPTION


async def test_sasl_client_to_plaintext_server(plaintext_bootstrap_servers):
    """Verify that a SASL-configured client fails when connecting to a plaintext server."""
    config = fluss.Config({
        "bootstrap.servers": plaintext_bootstrap_servers,
        "security.protocol": "sasl",
        "security.sasl.mechanism": "PLAIN",
        "security.sasl.username": "admin",
        "security.sasl.password": "admin-secret",
    })
    with pytest.raises(fluss.FlussError):
        await fluss.FlussConnection.create(config)

File: fluss-rust/bindings/cpp/test/test_sasl_auth.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include <gtest/gtest.h>

#include "test_utils.h"

class SaslAuthTest : public ::testing::Test {
   protected:
    const std::string& sasl_servers() {
        return fluss_test::FlussTestEnvironment::Instance()->GetSaslBootstrapServers();
    }
    const std::string& plaintext_servers() {
        return fluss_test::FlussTestEnvironment::Instance()->GetBootstrapServers();
    }
};

TEST_F(SaslAuthTest, SaslConnectWithValidCredentials) {
    fluss::Configuration config;
    config.bootstrap_servers = sasl_servers();
    config.security_protocol = "sasl";
    config.security_sasl_mechanism = "PLAIN";
    config.security_sasl_username = "admin";
    config.security_sasl_password = "admin-secret";

    fluss::Connection conn;
    ASSERT_OK(fluss::Connection::Create(config, conn));

    fluss::Admin admin;
    ASSERT_OK(conn.GetAdmin(admin));

    // Perform a basic operation to confirm the connection is fully functional
    std::string db_name = "cpp_sasl_test_valid_db";
    fluss::DatabaseDescriptor descriptor;
    descriptor.comment = "created via SASL auth";
    ASSERT_OK(admin.CreateDatabase(db_name, descriptor, true));

    bool exists = false;
    ASSERT_OK(admin.DatabaseExists(db_name, exists));
    ASSERT_TRUE(exists);

    ASSERT_OK(admin.DropDatabase(db_name, true, true));
}

TEST_F(SaslAuthTest, SaslConnectWithSecondUser) {
    fluss::Configuration config;
    config.bootstrap_servers = sasl_servers();
    config.security_protocol = "sasl";
    config.security_sasl_mechanism = "PLAIN";
    config.security_sasl_username = "alice";
    config.security_sasl_password = "alice-secret";

    fluss::Connection conn;
    ASSERT_OK(fluss::Connection::Create(config, conn));

    fluss::Admin admin;
    ASSERT_OK(conn.GetAdmin(admin));

    // Basic operation to confirm functional connection
    bool exists = false;
    ASSERT_OK(admin.DatabaseExists("some_nonexistent_db_alice", exists));
    ASSERT_FALSE(exists);
}

TEST_F(SaslAuthTest, SaslConnectWithWrongPassword) {
    fluss::Configuration config;
    config.bootstrap_servers = sasl_servers();
    config.security_protocol = "sasl";
    config.security_sasl_mechanism = "PLAIN";
    config.security_sasl_username = "admin";
    config.security_sasl_password = "wrong-password";

    fluss::Connection conn;
    auto result = fluss::Connection::Create(config, conn);
    ASSERT_FALSE(result.Ok());
    EXPECT_EQ(result.error_code, fluss::ErrorCode::AUTHENTICATE_EXCEPTION);
    EXPECT_NE(result.error_message.find("Authentication failed"), std::string::npos)
        << "Expected 'Authentication failed' in: " << result.error_message;
}

TEST_F(SaslAuthTest, SaslConnectWithUnknownUser) {
    fluss::Configuration config;
    config.bootstrap_servers = sasl_servers();
    config.security_protocol = "sasl";
    config.security_sasl_mechanism = "PLAIN";
    config.security_sasl_username = "nonexistent_user";
    config.security_sasl_password = "some-password";

    fluss::Connection conn;
    auto result = fluss::Connection::Create(config, conn);
    ASSERT_FALSE(result.Ok());
    EXPECT_EQ(result.error_code, fluss::ErrorCode::AUTHENTICATE_EXCEPTION);
    EXPECT_NE(result.error_message.find("Authentication failed"), std::string::npos)
        << "Expected 'Authentication failed' in: " << result.error_message;
}

TEST_F(SaslAuthTest, SaslClientToPlaintextServer) {
    fluss::Configuration config;
    config.bootstrap_servers = plaintext_servers();
    config.security_protocol = "sasl";
    config.security_sasl_mechanism = "PLAIN";
    config.security_sasl_username = "admin";
    config.security_sasl_password = "admin-secret";

    fluss::Connection conn;
    auto result = fluss::Connection::Create(config, conn);
    ASSERT_FALSE(result.Ok()) << "SASL client connecting to plaintext server should fail";
}
